airbyte-integrations/connectors/source-bing-ads/source_bing_ads/reports.py

```diff
@@ -0,0 +1,252 @@
+        if not self.aggregation_disabled:
+            report_request.Aggregation = self.client.report_aggregation
+
+        report_request.ExcludeColumnHeaders = exclude_column_headers
+        report_request.ExcludeReportFooter = exclude_report_footer
+        report_request.ExcludeReportHeader = exclude_report_header
+        report_request.Format = report_file_format
+        report_request.FormatVersion = "2.0"
+        report_request.ReturnOnlyCompleteData = return_only_complete_data
+        report_request.Time = time
+        report_request.ReportName = self.report_name
+        # Defines the set of accounts and campaigns to include in the report.
+        scope = reporting_service.factory.create("AccountThroughCampaignReportScope")
+        scope.AccountIds = {"long": [account_id]}
+        scope.Campaigns = None
+        report_request.Scope = scope
+
+        columns = reporting_service.factory.create(f"ArrayOf{self.report_name}Column")
+        getattr(columns, f"{self.report_name}Column").append(self.report_columns)
+        report_request.Columns = columns
+        return report_request
+
+    def parse_response(self, response: sudsobject.Object, **kwargs: Mapping[str, Any]) -> Iterable[Mapping]:
+        if response is not None:
+            for row in response.report_records:
+                yield {column: self.get_column_value(row, column) for column in self.report_columns}
+
+        yield from []
+
+    def get_column_value(self, row: _RowReportRecord, column: str) -> Union[str, None, int, float]:
+        """
+        Reads field value from row and transforms string type field to numeric if possible
+        """
+        value = row.value(column)
+        if value == "":
+            return None
+
+        if value is not None and column in REPORT_FIELD_TYPES:
+            if REPORT_FIELD_TYPES[column] == "integer":
+                value = 0 if value == "--" else int(value.replace(",", ""))
+            elif REPORT_FIELD_TYPES[column] == "number":
+                value = 0.0 if value == "--" else float(value.replace("%", "").replace(",", ""))
+
+        return value
+
+    def get_report_record_timestamp(self, datestring: str) -> int:
+        """
+        Parse report date field based on aggregation type
+        """
+        if self.aggregation_disabled:
+            date = pendulum.from_format(datestring, "M/D/YYYY")
+        else:
+            if self.client.report_aggregation == "Hourly":
+                date = pendulum.from_format(datestring, "YYYY-MM-DD|H")
+            else:
+                date = pendulum.parse(datestring)
+
+        return date.int_timestamp
+
+    def stream_slices(
+        self,
+        **kwargs: Mapping[str, Any],
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        for account in source_bing_ads.source.Accounts(self.client, self.config).read_records(SyncMode.full_refresh):
+            yield {"account_id": account["Id"]}
```

Comment:
I'm assuming this isn't in any sort of time-order using account_id as the slicing of stream slices but you're saving state for each account ID. 

I've discussed this with @keu recently and my main concern with this approach is that we have a potentially boundless state size. Is there a hard limit on the amount of account IDs there can be or could it potentially be millions?

Reply:
I think there should be strict limit of amount of accounts per user 


https://help.ads.microsoft.com/#apex/ads/en/52001/0
![2021-09-07_11-57](https://user-images.githubusercontent.com/5635747/132316364-833ead85-c0ea-4816-a9d6-27d6f0e8dcdf.png)